refactor!: enshrining owner in notes and state vars

benesjan opened the pull request:
Fixes F-33
Fixes F-179
Fixes https://github.com/AztecProtocol/aztec-packages/issues/15568

In this PR I enshrine owner in notes and in state variables dealing with notes. A motivation for this has been the following:
1. Privacy leak when initializing state variables: https://github.com/AztecProtocol/aztec-packages/issues/15568 (will be tackled in a followup PR by including owner nsk in the nullifier preimage),
2. notes always having a logical owner - we always need the information of an owner to be able to compute a nullifier in a way that the note sender is not able to derive (and hence have "spending" privacy from sender),
3. `PrivateSet` state var has been a bit of a lie as generally we always only worked with a subset of the set - subset of owner's notes (this subset got selected by deriving the storage slot based on the owner address via a `Map`),
4. this also enables stealth addresses as PXE will be able to know the semantic owner and return notes that belong to sub-addresses as well.

I didn't want to tackle https://github.com/AztecProtocol/aztec-packages/issues/15568 in this PR but the nature of the state variables changed and `PrivateMutable` and `PrivateImmutable` became scoped per user which requires them to be intializable for each user. With the original initialization nullifier scheme this didn't work as there was only 1 nullifier per the whole `PrivateMutable`. Introducing `PrivateMutable` and `PrivateImmutable` that can have only 1 owner (the contract admin) is expected to be done in a followup work.

## Issues with this PR
- Lack of these admin state vars make parts of the code quite weird (e.g. account contracts where by definition we have only 1 PublicKeyNote per contract but we access it via `private_immutable.at(self.address).get_note()`).
- State vars naming is now inconsistent: `PublicMutable` contains one value per the whole contract while `PrivateMutable` contains one value per contract-user pair (same goes for `PublicImmutable` and `PrivateImmutable`).

benesjan commented:
* **#18619** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18619?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18619?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on noir-projects/aztec-nr/aztec/src/messages/discovery/partial_notes.nr:
```diff
@@ -174,19 +178,22 @@ fn decode_partial_note_private_msg(
 
     assert(
-        msg_content.len() > PARTIAL_NOTE_PRIVATE_MSG_CONTENT_NON_NOTE_FIELDS_LEN,
+        msg_content.len() >= PARTIAL_NOTE_PRIVATE_MSG_CONTENT_NON_NOTE_FIELDS_LEN,
```
Now the private message commonly doesn't have any non-note fields.

nventuro replied:
I don't understand this change - if `msg_content == NON_NOTE_FIELDS`, does that not mean that the note has no content? ie

```noir
#[note]
struct EmptyNote {}
```

benesjan replied:
It does not because here we refer to message content and not note content and message content always needs to contain these fields: owner, storage slot, randomness, note_completion_log_tag.

Docs were lacking here so I added a comment in https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/2fdb409eb315edf8ce89788b2cfb3202685cd67b

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_immutable.nr:
```diff
@@ -41,22 +47,17 @@ use crate::state_vars::storage::HasStorageSlot;
+/// constant for the lifetime of the contract. To ensure there is only one configuration
+/// per contract you would consider valid only admin's view of the PrivateImmutable
+/// (i.e you would use `PrivateImmutable::at(...)` only with the admin's address in your
+/// contract).
```
Having this described here feels strange but it will hopefully go away with introduction of something like `AdminPrivateImmutable`.

nventuro replied:
Instead of using the admin address, wouldn't you use the contract address? The admin (owner of the note) might change, but the contract still has a single `PrivateImmutable`.

benesjan replied:
Seems like you have thought much more about this than me. Is the general idea that we would have the contract address on the input when constructing the state var and then have some standardized API of how to obtain an admin?

Updated the comment in https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/497d6ebe4f3315e226bc9d27335f3a9cd1967106 + created [this issue](https://linear.app/aztec-labs/issue/F-187/introduce-admin-only-state-variables) to track the design and implementation of admin state vars

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_immutable.nr:
```diff
@@ -90,55 +91,67 @@ impl<Note, Context> PrivateImmutable<Note, Context> {
     /// the storage location.
     ///
-    pub fn compute_initialization_nullifier(self) -> Field {
+    fn compute_initialization_nullifier(self) -> Field {
```
I had to move this function from the generic context impl. to the context specific ones because now the impls differ between `PrivateContext` and `UtilityContext`.

nventuro replied:
What if we kept the generic impl but changed it so that it receives the `secret` as a param? There'd be less duplication and we'd more clearly show the difference (oracle vs kernel req)

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/57ff8396cb450d43c7d7cd7473458954df2b4b77

benesjan commented on noir-projects/aztec-nr/uint-note/src/uint_note.nr:
```diff
@@ -50,9 +48,8 @@ impl NoteHash for UintNote {
         // perform both steps of the partial note computation.
 
         // First we create the partial note from a commitment to the private content (including storage slot).
-        let private_content = UintPartialNotePrivateContent { owner: self.owner };
```
The `UintPartialNotePrivateContent` was now an empty struct so I dropped and now there is just a standalone `compute_partial_commitment` function.

benesjan commented on noir-projects/noir-contracts/contracts/app/nft_contract/src/types/nft_note.nr:
```diff
@@ -315,13 +298,10 @@ mod test {
             subarray(partial_note.compute_note_completion_log(token_id), 1);
 
-        assert_eq(private_log_without_public_tag.concat(public_log_without_tag), note.pack());
+        assert_eq(public_log_without_tag, note.pack());
```
This got simplified because now the note definition doesn't contain any private fields

benesjan commented on noir-projects/noir-contracts/contracts/test/spam_contract/src/types/balance_set.nr:
```diff
@@ -1,57 +1,64 @@
-// This file is copied from the token contract.
```
The change here is large because this was a copy from the token contract but at some older time. I couldn't be bothered updating multiple versions of the same thing so I just copied the `balance_set.nr` from the token contract here.

benesjan commented on noir-projects/noir-contracts/contracts/test/spam_contract/src/types/token_note.nr:
```diff
@@ -1,24 +0,0 @@
-use aztec::{macros::notes::note, protocol_types::{address::AztecAddress, traits::Packable}};
```
The `BalanceSet` from token contract uses normal `UintNote` instead of this custom non-sense

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable/test.nr:
```diff
@@ -1,45 +1,56 @@
-use crate::{context::{PrivateContext, UtilityContext}, state_vars::private_mutable::PrivateMutable};
```
Now we include `nsk_app` in the initialization nullifiers so I needed to create accounts here as the `utilityGetKeyValidationRequest` is called while getting the secret.

nventuro commented on boxes/boxes/vite/src/contracts/src/main.nr:
```diff
@@ -6,44 +6,44 @@ contract BoxReact {
+        let new_number = ValueNote::new(number);
 
-        numbers.at(owner).initialize(new_number).emit(
+        owner_number.initialize(new_number).emit(
```
Surprised the unnecessary `mut` was not raising warnings.

benesjan replied:
I think we might have stricter compilation params in `/noir-projects`

nventuro commented on boxes/boxes/vite/src/contracts/src/main.nr:
```diff
@@ -6,44 +6,44 @@ contract BoxReact {
-            owner,
-            MessageDelivery.CONSTRAINED_ONCHAIN,
-        );
+        let owner_number = self.storage.numbers.at(owner);
```
Unrelated to this PR, but I'd just inline all of these.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/eaacf662a6cc7cc4ed91d736e2de845601cb3696

nventuro commented on docs/docs/developers/docs/resources/migration_notes.md:
```diff
@@ -9,6 +9,118 @@ Aztec is in full-speed development. Literally every version breaks compatibility
 
 ## TBD
 
+### [Aztec.nr] Note owner is now enshrined
```
Very nice

nventuro commented on docs/examples/contracts/bob_token_contract/src/main.nr:
```diff
@@ -100,16 +100,16 @@ pub contract BobToken {
         let sender = self.msg_sender().unwrap();
         // Spend sender's notes (consumes existing notes)
-        self.storage.private_balances.at(sender).sub(amount, sender);
+        self.storage.private_balances.sub(amount, sender);
```
Brings tears to my eyes.

nventuro commented on docs/examples/tutorials/token_bridge_contract/contracts/aztec/nft/src/nft.nr:
```diff
@@ -11,17 +11,14 @@ use dep::aztec::{
 // docs:start:nft_note_new
 impl NFTNote {
-    pub fn new(owner: AztecAddress, token_id: Field) -> Self {
-        // The randomness preserves privacy by preventing brute-forcing
-        NFTNote { owner, randomness: unsafe { random() }, token_id }
```
@mverzilli you missed this one 😛

nventuro commented on noir-projects/aztec-nr/aztec/src/messages/discovery/partial_notes.nr:
```diff
@@ -15,14 +15,15 @@ use protocol_types::{
-global PARTIAL_NOTE_PRIVATE_MSG_CONTENT_STORAGE_SLOT_INDEX: u32 = 0;
-global PARTIAL_NOTE_PRIVATE_MSG_CONTENT_RANDOMNESS_INDEX: u32 = 1;
-global PARTIAL_NOTE_PRIVATE_MSG_CONTENT_NOTE_COMPLETION_LOG_TAG_INDEX: u32 = 2;
+/// owner, storage slot, randomness, note_completion_log_tag
```
```suggestion
/// [ owner, storage slot, randomness, note_completion_log_tag ]
```

nventuro commented on noir-projects/aztec-nr/aztec/src/messages/processing/mod.nr:
```diff
@@ -70,10 +70,16 @@ pub(crate) unconstrained fn get_private_logs(
-/// `recipient` is the account to which the note was sent to. Other accounts will not be able to access this note (e.g.
-/// other accounts will not be able to see one another's token balance notes, even in the same PXE) unless authorized.
+/// `owner` is the address used in note hash and nullifier computation. Only this address has the keys required to
+/// nullify (spend) the note.
```
```suggestion
/// `owner` is the address used in note hash and nullifier computation, often requiring knowledge of their
/// nullifier secret key.
```

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable/test.nr:
```diff
@@ -1,45 +1,56 @@
-    PrivateMutable::new(context, STORAGE_SLOT)
+    owner: AztecAddress,
+) -> OwnedPrivateMutable<MockNote, &mut PrivateContext> {
+    PrivateMutable::new(context, STORAGE_SLOT).at(owner)
```
I'd not change these helpers, would instead return `PrivateMutable` and  do the `at` in each test. That more clearly mirrors what happens in a contract as storage is accessed.

benesjan replied:
Makes sense https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/4deef54adc757b8154e1b83d6ba250d3ce046007

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/private_set/test.nr:
```diff
@@ -1,21 +1,23 @@
     context: &mut PrivateContext,
-) -> PrivateSet<MockNote, &mut PrivateContext> {
-    PrivateSet::new(context, STORAGE_SLOT)
+) -> OwnedPrivateSet<MockNote, &mut PrivateContext> {
```
Same

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/4deef54adc757b8154e1b83d6ba250d3ce046007

nventuro commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -159,55 +174,25 @@ impl<Note, Context> PrivateMutable<Note, Context> {
+    fn compute_initialization_nullifier(self) -> Field {
+        let owner_npk_m = get_public_keys(self.owner).npk_m;
+        let owner_npk_m_hash = owner_npk_m.hash();
+        let secret = self.context.request_nsk_app(owner_npk_m_hash);
```
Same

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/57ff8396cb450d43c7d7cd7473458954df2b4b77

nventuro commented on noir-projects/aztec-nr/aztec/src/test/mocks/mock_note.nr:
```diff
@@ -27,14 +27,20 @@ impl NoteType for MockNote {
     fn compute_nullifier(
         _self: Self,
         _context: &mut PrivateContext,
+        _owner: AztecAddress,
```
Why not use the owner now that we have it? We could even use the `#[note]` macro. We'd need to use `env.create_light_account` though.

benesjan replied:
Using the `#[note]` macro there doesn't work:

<img width="502" height="208" alt="image" src="https://github.com/user-attachments/assets/d105b933-0e7e-4385-b673-194d42584278" />

Given that it's just a MockNote used in unit tests I feel like this change is not worth the extra call to `env.create_light_account`.

nventuro commented on noir-projects/aztec-nr/uint-note/src/uint_note.nr:
```diff
@@ -27,19 +27,17 @@ use dep::aztec::{
 #[derive(Deserialize, Eq, Serialize, Packable)]
 #[custom_note]
 pub struct UintNote {
-    // The ordering of these fields is important given that it must:
-    //   a) match that of UintPartialNotePrivateContent, and
-    //   b) have the public field at the end
-    // Correct ordering is checked by the tests in this module.
-
-    /// The owner of the note, i.e. the account whose nullifier secret key is required to compute the nullifier.
```
I'm so happy

nventuro commented on noir-projects/noir-contracts/contracts/app/claim_contract/src/main.nr:
```diff
@@ -34,24 +34,20 @@ pub contract Claim {
             target_address == proof_retrieved_note.contract_address,
             "Note does not correspond to the target contract",
         );
-        assert_eq(owner, self.msg_sender().unwrap(), "Note does not belong to the sender");
```
I'm so happy

nventuro commented on yarn-project/pxe/src/contract_function_simulator/oracle/private_execution_oracle.ts:
```diff
@@ -311,16 +313,19 @@ export class PrivateExecutionOracle extends UtilityExecutionOracle implements IP
     );
-    const dbNotesFiltered = dbNotes.filter(n => !pendingNullifiers.has((n.siloedNullifier as Fr).value));
+    const dbNotesFiltered = dbNotes.filter(
+      n => !pendingNullifiers.has((n.siloedNullifier as Fr).value) && n.owner.equals(owner),
```
Why do we need to check for the owner? `dbNotes` was retrieved by filtering by owner already, so it seems like this would always be true.

benesjan replied:
This is a mess I left behind from previous WIP versions. Good catch! Thanks

https://github.com/AztecProtocol/aztec-packages/pull/18619/commits/9dab0628aed20cc196fb9e6b4127662a916f97f3

nventuro commented on yarn-project/stdlib/src/note/notes_filter.ts:
```diff
@@ -16,6 +16,8 @@ export type NotesFilter = {
    * @remarks Providing a contract address is required as we need that information to trigger private state sync.
    */
   contractAddress: AztecAddress;
+  /** The owner of the note. */
+  owner?: AztecAddress;
```
Should this not be mandatory?

benesjan replied:
Good question. I don't think it's necessary for it to be mandatory because:
1.  It makes the [getNotes test wallet API](https://github.com/AztecProtocol/aztec-packages/blob/b0f3eedd8578b02d7d03a0f2e1dba1f0af437f67/yarn-project/test-wallet/src/wallet/test_wallet.ts#L253) more useful for debugging,
2. we already have storage slot as optional and it feels like the owner has the same importance (this is not an argument for correctness of the decision but it would be an argument for tackling this in a separate PR).

But I would say it might make sense to make it mandatory along with storage slot as it feels like we should not be making decisions based on the test wallet debugging utility.

If you agree I would do this in a separate PR as making it mandatory would change quite a few tests.

nventuro reviewed (approved):
This PR just makes so much sense. Thank you.

benesjan committed:
refactor!: enshrining owner in notes and state vars

Fixes F-33
Fixes F-179
Fixes https://github.com/AztecProtocol/aztec-packages/issues/15568

In this PR I enshrine owner in notes and in state variables dealing with notes. A motivation for this has been the following:
1. Privacy leak when initializing state variables: https://github.com/AztecProtocol/aztec-packages/issues/15568 (will be tackled in a followup PR by including owner nsk in the nullifier preimage),
2. notes always having a logical owner - we always need the information of an owner to be able to compute a nullifier in a way that the note sender is not able to derive (and hence have "spending" privacy from sender),
3. `PrivateSet` state var has been a bit of a lie as generally we always only worked with a subset of the set - subset of owner's notes (this subset got selected by deriving the storage slot based on the owner address via a `Map`),
4. this also enables stealth addresses as PXE will be able to know the semantic owner and return notes that belong to sub-addresses as well.

I didn't want to tackle https://github.com/AztecProtocol/aztec-packages/issues/15568 in this PR but the nature of the state variables changed and `PrivateMutable` and `PrivateImmutable` became scoped per user which requires them to be intializable for each user. With the original initialization nullifier scheme this didn't work as there was only 1 nullifier per the whole `PrivateMutable`. Introducing `PrivateMutable` and `PrivateImmutable` that can have only 1 owner (the contract admin) is expected to be done in a followup work.

## Issues with this PR
- Lack of these admin state vars make parts of the code quite weird (e.g. account contracts where by definition we have only 1 PublicKeyNote per contract but we access it via `private_immutable.at(self.address).get_note()`).
- State vars naming is now inconsistent: `PublicMutable` contains one value per the whole contract while `PrivateMutable` contains one value per contract-user pair (same goes for `PublicImmutable` and `PrivateImmutable`).
